rfcs/0133-git-hashing.md

```diff
@@ -0,0 +1,346 @@
+   RFCs like this show that the so-far more sleepy lower layers also have plenty of potential to gain sophistication too.
+
+   Embracing layering on technical, educational, communications, and managerial levels can scale our capacity to manage complexity and sophistication without the project growing out of control.
+   It will "divide and conquer" the project so the interfaces between each layer are still rigorously enforced preventing a combinatorial explosion in complexity.
+   That frees up "complexity budget" for project like this.
+
+   We plan on more formally proposing this next.
+
+## Git and Nix's file system data models do not entirely coincide
+
+Nix puts the permission info of a file (executable bit for now) with that file, whereas Git puts it with the name and hash in the directory.
+The practical effect of this discrepancy is that a root file (as opposed to directory) in Nix has permission info, but does not in Git.
+
+If we are trying to convert existing Nix data into Git, this is a problem.
+Assuming we treat "no permission bits" as meaning "non-executable", we will have a partial conversion that will fail on executable files without a parent directory.
+Tricks like always wrapping everything in a directory get around this, but then we have to be careful the directory is exactly as expected when "unwrapping" in the other direction.
+
+For now, we only focus on ingesting data *from* Git *to* Nix, and this side-steps the issue.
+That conversation is total (though not surjective), and so there is no problem for now.
+
+# Alternatives
+[alternatives]: #alternatives
+
+The dependency graph of steps can be sliced to save some for future work.
+For now they are all written together, but during the RFC meetings we will decide which steps (if any) to ratify now, and which steps to save for later.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+None at this time.
+
+# Future work --- Software Heritage
+[future]: #future-work-swh
+
+## Motivation
+
+Software Heritage is a great way to get source code that is no longer found at its original location online.
+It would be really great to use it as a substituter to fetch source code that fails to be fetched the default way.
+
+## Detailed Designed
+
+- **Depends on**: Nix-agnostic content-addressing "stores"
+
+[`GET /api/1/raw/(swhid)/`](https://docs.softwareheritage.org/devel/swh-web/uri-scheme-api.html#get--api-1-raw-(swhid)-)
+is a relatively new Software Heritage API endpoint to get a git object for the given SWHID.
+We can used this to write a Software Heritage store, which will accomplish the task layed out in the previous motivation section.
+
+# Future work --- IPFS
+[future]: #future-work-ipfs
+
+## Motivation
+
+### Binary Distribution
+
+IPFS is a potential solution leveraging content addressing for a P2P CDN.
+It out-of-the-box match's Nix's data model extremely well.
```

Comment:
This sentence doesn't read properly for me. Did you mean to say "It's out-of-the box data model matches Nix's data model very well."?

Reply:
I think I meant
```suggestion
It, out-of-the-box, matches Nix's data model extremely well.
```

Reply:
Ok that works for me. However I think that it is easier to read like this

```suggestion
Out-of-the-box it matches Nix's data model extremely well.
```

Reply:
Sure, though I think I'll put back in the single comma :)